File: code/haskell/exercises.hs
module Exercises
  ( firstThenApply
  , powers
  , meaningfulLineCount
  , volume
  , surfaceArea
  , insert
  , contains
  , size
  , inorder
  , change
  , Shape(..)
  , BST(Empty)
  ) where

import Data.Char (isSpace)
import Data.List (find)
import Data.Map (Map)
import qualified Data.Map.Strict as M

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs p f = f <$> find p xs

powers :: Integral a => a -> [a]
powers base = iterate ((* base)) 1

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
  contents <- readFile path
  let ls :: [String]
      ls = lines contents
      isMeaningful :: String -> Bool
      isMeaningful s =
        case dropWhile isSpace s of
          ""     -> False
          '#':_  -> False
          _      -> True
  pure (length (filter isMeaningful ls))

data Shape
  = Sphere Double
  | Box Double Double Double
  deriving (Eq)

instance Show Shape where
  show (Sphere r)  = "Sphere " ++ show r
  show (Box w l d) = "Box "    ++ show w ++ " " ++ show l ++ " " ++ show d

volume :: Shape -> Double
volume (Sphere r)  = (4 / 3) * pi * r * r * r
volume (Box w l d) = w * l * d

surfaceArea :: Shape -> Double
surfaceArea (Sphere r)  = 4 * pi * r * r
surfaceArea (Box w l d) = 2 * (w * l + w * d + l * d)

data BST a
  = Empty
  | Node (BST a) a (BST a)

instance Show a => Show (BST a) where
  show Empty = "()"
  show (Node l v r) =
    "(" ++ leftPart ++ show v ++ rightPart ++ ")"
    where
      leftPart  = case l of
                    Empty -> ""
                    _     -> show l
      rightPart = case r of
                    Empty -> ""
                    _     -> show r

insert :: Ord a => a -> BST a -> BST a
insert x Empty = Node Empty x Empty
insert x (Node l v r)
  | x == v    = Node l v r
  | x <  v    = Node (insert x l) v r
  | otherwise = Node l v (insert x r)

contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node l v r)
  | x == v    = True
  | x <  v    = contains x l
  | otherwise = contains x r

size :: BST a -> Int
size Empty        = 0
size (Node l _ r) = 1 + size l + size r

inorder :: BST a -> [a]
inorder Empty        = []
inorder (Node l v r) = inorder l ++ [v] ++ inorder r

change :: Integer -> Either String (Map Integer Integer)
change amount
  | amount < 0 = Left "amount must be nonnegative"
  | otherwise  =
      let (q, r1) = amount `divMod` 25
          (d, r2) = r1     `divMod` 10
          (n, p)  = r2     `divMod` 5
      in Right (M.fromList [(25, q), (10, d), (5, n), (1, p)])

File: code/typescript/exercises.ts
import * as fs from "fs";

export function firstThenApply<A, B>(
  arr: readonly A[],
  predicate: (x: A) => boolean,
  fn: (x: A) => B
): B | undefined {
  for (const x of arr) {
    if (predicate(x)) return fn(x);
  }
  return undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint, never, void> {
  let current: bigint = 1n;
  while (true) {
    yield current;
    current = current * base;
  }
}

export async function meaningfulLineCount(path: string): Promise<number> {
  const stream: fs.ReadStream = fs.createReadStream(path, { encoding: "utf8" });
  let buffer: string = "";
  let count: number = 0;

  return await new Promise<number>((resolve, reject) => {
    stream.on("data", (chunk: string | Buffer) => {
      buffer += typeof chunk === "string" ? chunk : chunk.toString("utf8");
      const lines: string[] = buffer.split(/\r?\n/);
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        const trimmed: string = line.trim();
        if (trimmed !== "" && !trimmed.startsWith("#")) count += 1;
      }
    });

    stream.on("end", () => {
      const trimmed: string = buffer.trim();
      if (trimmed !== "" && !trimmed.startsWith("#")) count += 1;
      resolve(count);
    });

    stream.on("error", (err: Error) => reject(err));
  });
}

export type Sphere = Readonly<{ kind: "Sphere"; radius: number }>;
export type Box = Readonly<{ kind: "Box"; width: number; length: number; depth: number }>;
export type Shape = Sphere | Box;

export function volume(s: Shape): number {
  switch (s.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * s.radius * s.radius * s.radius;
    case "Box":
      return s.width * s.length * s.depth;
  }
}

export function surfaceArea(s: Shape): number {
  switch (s.kind) {
    case "Sphere":
      return 4 * Math.PI * s.radius * s.radius;
    case "Box": {
      const { width, length, depth } = s;
      return 2 * (width * length + width * depth + length * depth);
    }
  }
}

type Comparable = string | number | boolean;

export abstract class BinarySearchTree<T extends Comparable> {
  public abstract insert(value: T): BinarySearchTree<T>;
  public abstract contains(value: T): boolean;
  public abstract size(): number;
  public abstract inorder(): Generator<T, void, unknown>;
  public abstract toString(): string;
}

export class Empty<T extends Comparable> extends BinarySearchTree<T> {
  public insert(value: T): BinarySearchTree<T> {
    return new Node<T>(value, new Empty<T>(), new Empty<T>());
  }
  public contains(_value: T): boolean {
    return false;
  }
  public size(): number {
    return 0;
  }
  public *inorder(): Generator<T, void, unknown> {}
  public toString(): string {
    return "()";
  }
}

class Node<T extends Comparable> extends BinarySearchTree<T> {
  private readonly value: T;
  private readonly left: BinarySearchTree<T>;
  private readonly right: BinarySearchTree<T>;

  public constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    super();
    this.value = value;
    this.left = left;
    this.right = right;
  }

  private toComparable(v: T): number | string {
    return typeof v === "boolean" ? Number(v) : v;
  }

  public insert(value: T): BinarySearchTree<T> {
    if (value === this.value) return this; 
    if (this.toComparable(value) < this.toComparable(this.value)) {
      return new Node<T>(this.value, this.left.insert(value), this.right);
    } else {
      return new Node<T>(this.value, this.left, this.right.insert(value));
    }
  }

  public contains(value: T): boolean {
    if (value === this.value) return true;
    if (this.toComparable(value) < this.toComparable(this.value)) {
      return this.left.contains(value);
    }
    return this.right.contains(value);
  }

  public size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  public *inorder(): Generator<T, void, unknown> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  public toString(): string {
    const leftStr: string = this.left.toString();
    const rightStr: string = this.right.toString();
    const leftPart: string = leftStr === "()" ? "" : leftStr;
    const rightPart: string = rightStr === "()" ? "" : rightStr;
    return `(${leftPart}${this.value}${rightPart})`;
  }
}
